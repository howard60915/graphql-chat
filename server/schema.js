const schema = `
input RegisterByEmailMutationInput {
  email: String!
  password: String!
}

type RegisterByEmailMutationPayload {
  status: ResponseStatus
  user: User
}

input UpdateUserMutationInput {
  userId: String!
  email: String
}

type UpdateUserMutationPayload {
  status: ResponseStatus
  user: User
}

input DelUserMutationInput {
  userId: String!
}

type DelUserMutationPayload {
  status: ResponseStatus
  user: User
}

input AddConversationMutationInput {
  name: String
}

type AddConversationMutationPayload {
  status: ResponseStatus
  conversation: Conversation
}

input AddMessageMutationInput {
  conversationId: String!
  content: String!
}

type AddMessageMutationPayload {
  status: ResponseStatus
  message: Message
}

type Mutation {
  registerByEmail(input: RegisterByEmailMutationInput!): RegisterByEmailMutationPayload
  updateUser(input: UpdateUserMutationInput!): UpdateUserMutationPayload
  delUser(input: DelUserMutationInput!): DelUserMutationPayload
  addConversation(input: AddConversationMutationInput!): AddConversationMutationPayload
  addMessage(input: AddMessageMutationInput!): AddMessageMutationPayload
}

type Query {
  users: [User]
  conversations: [Conversation]
  messages: [Message]
}

type Subscription {
  onNewMessage: Message
}

enum ResponseStatus {
  failed
  ok
}

type User {
  id: String
  email: String
  isAdmin: Boolean
}

type Conversation {
  id: String
  name: String
}

type Message {
  id: String
  user: User
  conversation: Conversation
  content: String
}
`;

export default schema;
