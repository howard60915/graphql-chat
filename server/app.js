import graphqlServer from "./graphql";

const PORT = process.env.NODE_PORT || process.env.PORT || 8080;

const options = {
  port: PORT,
  endpoint: "/graphql",
  playground: "/playground"
};

graphqlServer.start(options, ({ port }) => {
  // eslint-disable-next-line no-console
  console.log(`> Ready on http://localhost:${port}`);
});
