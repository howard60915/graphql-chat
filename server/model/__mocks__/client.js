import { Client } from "pg";
import { mock } from "graphql-shortcake";
import knex from "knex";
import config from "../../../knexfile";

const { POSTGRES_HOST } = process.env;

const client = jest.requireActual("../client").default;

const db = knex(config);

const mockClient = mock(client);

mockClient.reset = async name => {
  const pgClient = new Client({
    host: POSTGRES_HOST || "127.0.0.1",
    user: "postgres",
    password: null,
    database: "postgres"
  });
  await pgClient.connect();
  await pgClient.query(`DROP DATABASE IF EXISTS ${name}`);
  await pgClient.query(`CREATE DATABASE ${name}`);
  await pgClient.end();
  await db.migrate.latest();
};

export default mockClient;
