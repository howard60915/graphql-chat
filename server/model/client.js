import { Client } from "graphql-shortcake";
import config from "../../knexfile";

export default new Client({ connectionString: config.connection });
