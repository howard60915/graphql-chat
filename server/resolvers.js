import UsersQuery from "./query/UsersQuery";
import RegisterByEmailMutation from "./mutation/RegisterByEmailMutation";
import UpdateUserMutation from "./mutation/UpdateUserMutation";
import DelUserMutation from "./mutation/DelUserMutation";
import AddMessageMutation from "./mutation/AddMessageMutation";

const Query = {
  users: UsersQuery
};

const Mutation = {
  registerByEmail: RegisterByEmailMutation,
  updateUser: UpdateUserMutation,
  delUser: DelUserMutation,
  addMessage: AddMessageMutation
};

export default { Query, Mutation };
