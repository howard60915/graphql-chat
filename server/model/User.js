import { GlobalModel, Hash } from "graphql-shortcake";
import client from "./client";

export default class User extends GlobalModel {
  static client = client;

  static hasOperator = true;

  static operator = () => User;

  static tableName = "user";

  static columns = {
    email: { type: String },
    password: { type: Hash, paths: ["archive"] },
    isAdmin: { type: Boolean, paths: ["archive"] }
  };

  async checkWritability(user, ThrowError) {
    const result = this.id === user.id || user.isAdmin;
    if (!result) throw new ThrowError();
    return result;
  }
}
