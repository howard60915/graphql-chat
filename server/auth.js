import { NotFoundError } from "graphql-schema-sushi";
import { AuthenticationError } from "apollo-server-errors";
import { JsonWebToken } from "nextjs-apollo/server";
import User from "./model/User";
import Certificate from "./model/Certificate";

const onSignIn = async data => {
  const { email, password, correlationId } = data;

  const { fetcher } = User;
  const user = await fetcher.where("email", email).fetch(NotFoundError);

  if (!(password && user.password.verify(password))) {
    throw new AuthenticationError("AuthenticationError");
  }

  return { user: user.previous, correlationId };
};

const onSignOut = () => ({});

const onRenew = async data => {
  const { user, correlationId } = data;

  const renewUser = await User.load(user.id, NotFoundError);

  return { user: renewUser.previous, correlationId };
};

const toUser = ({ user }) => {
  if (!user.id) throw new AuthenticationError();
  return User.forge(user);
};

const jwt = new JsonWebToken(Certificate, {
  onSignIn,
  onSignOut,
  onRenew,
  toUser
});

export default jwt;
