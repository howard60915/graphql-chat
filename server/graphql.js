import { GraphQLServer } from "graphql-yoga";
import auth from "./auth";

const server = new GraphQLServer({
  typeDefs: `
    type Query {
      hello: String
    }
  `,
  resolvers: {
    Query: {
      hello: () => "hello"
    }
  }
});

auth.applyMiddleware(server.express);

export default server;
