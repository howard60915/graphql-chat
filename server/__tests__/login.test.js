import axios from "axios";
import server from "../graphql";
import client from "../model/client";
import User from "../model/User";

const authURL = "http://127.0.0.1:4000/auth";

describe("auth", () => {
  let refreshToken;
  const correlationId = "00000000-0000-0000-0000-000000000100";

  beforeAll(() => server.start());

  afterAll(() => server.close());

  describe("sign in", () => {
    const body = { email: "howhow@howfun.co", password: "howhowno1" };
    const user = new User(body);

    const request = (data, id) =>
      axios.post(
        authURL,
        { ...data, method: "signIn" },
        { headers: { cookie: `x-correlation-id=${id || ""}` } }
      );

    it("successfully signIn", async () => {
      client.mockReturnValueOnce([
        {
          id: "00000000-0000-0000-0000-000000000010",
          email: "howhow@howfun.co",
          archive: { password: user.password.value }
        }
      ]);
      const { data } = await request(body, correlationId);
      expect(data.refreshToken).not.toBeUndefined();
      expect(data.accessToken).not.toBeUndefined();
      expect(data).toEqual(
        expect.objectContaining({
          user: expect.objectContaining({
            id: "00000000-0000-0000-0000-000000000010"
          })
        })
      );
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
      ({ refreshToken } = data);
    });

    it("signIn with wrong password", async () => {
      client.mockReturnValueOnce([
        {
          id: "00000000-0000-0000-0000-000000000010",
          email: "howhow@howfun.co",
          archive: { password: user.password.value }
        }
      ]);
      await expect(
        request({
          email: "howhow@howfun.co",
          password: "howhowsuck"
        })
      ).rejects.toEqual(new Error("Request failed with status code 403"));
    });
  });
  describe("sign out", () => {
    const request = (data, id) =>
      axios.post(
        authURL,
        {
          ...data,
          method: "signOut"
        },
        { headers: { cookie: `x-correlation-id=${id || ""}` } }
      );

    it("successfully signOut", async () => {
      const { headers, data } = await request({}, correlationId);
      expect(data.status).toBe("ok");
      expect(client).toHaveBeenCalledTimes(0);
      expect(headers["set-cookie"]).toBeUndefined();
    });
  });
  describe("renew", () => {
    const request = (data, id) =>
      axios.post(
        authURL,
        {
          ...data,
          method: "renew"
        },
        { headers: { cookie: `x-correlation-id=${id || ""}` } }
      );

    it("successfully renew", async () => {
      client.mockReturnValueOnce([
        {
          id: "00000000-0000-0000-0000-000000000010",
          email: "howhow@howfun.co"
        }
      ]);
      const { data } = await request({ refreshToken }, correlationId);
      expect(data.refreshToken).not.toBeUndefined();
      expect(data.accessToken).not.toBeUndefined();
      expect(data).toEqual(
        expect.objectContaining({
          user: expect.objectContaining({
            id: "00000000-0000-0000-0000-000000000010"
          })
        })
      );
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });

    it("user not found", async () => {
      await expect(request({ refreshToken })).rejects.toEqual(
        new Error("Request failed with status code 403")
      );
    });
  });
});
