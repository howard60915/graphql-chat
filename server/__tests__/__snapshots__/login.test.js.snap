// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth renew successfully renew 1`] = `
Array [
  "SELECT * FROM \\"user\\" WHERE \\"user\\".\\"deleted_at\\" IS NULL AND \\"id\\" IN ($1)",
]
`;

exports[`auth sign in successfully signIn 1`] = `
Array [
  "SELECT * FROM \\"user\\" WHERE \\"user\\".\\"deleted_at\\" IS NULL AND \\"email\\" = $1 LIMIT 1",
]
`;
